x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealm.java

```diff
@@ -423,17 +427,38 @@ public void authenticate(final AuthenticationToken authenticationToken, final Ac
                 jwt = SignedJWT.parse(serializedJwt.toString());
                 final String jwtAlg = jwt.getHeader().getAlgorithm().getName();
                 final boolean isJwtAlgHmac = JwtRealmSettings.SUPPORTED_SIGNATURE_ALGORITHMS_HMAC.contains(jwtAlg);
-                final JwtRealm.JwksAlgs jwksAndAlgs = isJwtAlgHmac ? this.jwksAlgsHmac : this.jwksAlgsPkc;
-                JwtValidateUtil.validate(
-                    jwt,
-                    this.allowedIssuer,
-                    this.allowedAudiences,
-                    this.allowedClockSkew.seconds(),
-                    jwksAndAlgs.algs,
-                    jwksAndAlgs.jwks
-                );
+
+                try:
+                    JwtRealm.JwksAlgs jwksAndAlgs = isJwtAlgHmac ? this.jwksAlgsHmac : this.jwksAlgsPkc;
+                    JwtValidateUtil.validate(
+                        jwt,
+                        this.allowedIssuer,
+                        this.allowedAudiences,
+                        this.allowedClockSkew.seconds(),
+                        jwksAndAlgs.algs,
+                        jwksAndAlgs.jwks
+                    );
+                } catch (JwtValidateUtil.InvalidSignatureException e) {
+                    /**
+                     * Elasticsearch automatically caches the retrieved JWK set to avoid unnecessary HTTP requests, 
+                     * but will attempt to refresh the JWK upon signature verification failure, 
+                     * as this might indicate that the JWT Provider has rotated the signing keys.
+                     */
+                    this.refreshJwksAlgs();
```

Comment:
Based on `isJwtAlgHmac`, only one JWKSet needs to be refreshed, not both PKC and HMAC.
1. If isJwtAlgHmac=false, only the PKC JWKSet needs refresh.
a. Get content from URL or file.
2. If isJwtAlgHmac=true, only the HMAC JWKSet needs refresh.
a. Get content from elasticsearch-keystore. Note, only HMAC JWKSet or HMAC OIDC-compatible JWK can be set at any time.

There are some corner cases where revalidation is not required. Validation uses CPU, and we generally want to minimize the impact on resources that might be in use for search queries or data ingest.

3. If there was an error during reload, keep old settings (ex: network connectivity issue connecting to PKC HTTPS URL, etc).
4. Compare old vs new JWKSet content. If exact match, there is no need to attempt revalidation.
5. If content changed, then set the corresponding this.jwkAlgsPkc or this.jwkAlgsHmac, and run this.verifyAnyAvailableJwkAndAlgPair.
6. If this.verifyAnyAvailableJwkAndAlgPair throws an error, skip validation. We detected the realm config is now in an unusable state. Intervention may be needed to fix the config (ex: PKC JWKSet changes JWK attributes from RS256=>RS512, ES256=>ES512, or HS256=>HS512, but the realm only allows RS256,ES256,HS256 signatures).
7. If this.verifyAnyAvailableJwkAndAlgPair is OK, then revalidation can be attempted.

Reply:
Agreed seems my original solution was a bit "quick and dirty". I have implemented your suggestions. One thing I was wondering though `Intervention may be needed to fix the config` do we completely throw exceptions or keep accepting JWTs with old signed keys? I assume nothing will happen with the cached JWT keys even if the new key is raising exceptions?

Reply:
I think we need to accept the new JWKs, and invalidate the cache to stop old JWTs from working.

That node is now in a degraded state. We might be stopping JWT authentication from continuing to work on that node. However, all of the other things will continue working using all other authentication methods, such as searching, indexing, shard replication, cross-cluster follower indexing, etc.

The sooner that JWT authentication problem is noticed, the sooner it can be investigated and fixed.

Consider what happens if a JWT config issue is left for too long. Nodes will start getting bounced as part of normal cloud providers recycling VMs. Those restarted nodes will fail to start. Now all functionality of that node is lost. That is a much more urgent situation. The operator has to do reactive fixing.

If we accept the new JWKs as-is, and invalidate the cache, there is time to report the problem and do proactive fixing of the JWT realm configuration.

Reply:
I agree with that logic, good to hear though that only the JWT Realm stops working. Was wondering that because I don't know the full inner workings of Elastic Search (yet).

It is indeed the responsibility of the user to fix the configuration or the JWTIssuer Source. So the most secure option for ES is to make this realm in an unusable state.

I have implemented said logic.